src/algorithm/contact-dynamics.hpp

```diff
@@ -72,6 +72,70 @@ namespace pinocchio
                                                 const Eigen::MatrixBase<ConstraintMatrixType> & J,
                                                 const Eigen::MatrixBase<KKTMatrixType> & MJtJ_inv);
   
+  ///
+  /// \brief Compute the impulse dynamics with contact constraints.
+  /// \note It computes the following problem: <BR>
+  ///       <CENTER> \f$ \begin{eqnarray} \underset{\dot{q}^{+}}{\min} & & \| \dot{q}^{+} - \dot{q}^{-} \|_{M(q)} \\
+  ///           \text{s.t.} & & J (q) \dot{q}^{+} = - \epsilon J (q) \dot{q}^{-}  \end{eqnarray} \f$ </CENTER> <BR>
+  ///       where \f$ \dot{q}^{-} \f$ is the generalized velocity before impact,
+  ///       \f$ M \f$ is the joint space mass matrix, \f$ J \f$ the constraint Jacobian and \f$ \epsilon \f$ is the coefficient of restitution (1 for a fully elastic impact or 0 for a rigid impact).
+  ///
+  /// \tparam JointCollection Collection of Joint types.
+  /// \tparam ConfigVectorType Type of the joint configuration vector.
+  /// \tparam TangentVectorType Type of the joint velocity vector.
+  /// \tparam ConstraintMatrixType Type of the constraint matrix.
+  ///
+  /// \param[in] model The model structure of the rigid body system.
+  /// \param[in] data The data structure of the rigid body system.
+  /// \param[in] q The joint configuration (vector dim model.nq).
+  /// \param[in] v_before The joint velocity before impact (vector dim model.nv).
+  /// \param[in] J The Jacobian of the constraints (dim nb_constraints*model.nv).
+  /// \param[in] inv_damping Damping factor for Cholesky decomposition of JMinvJt. Set to zero if constraints are full rank.
+  /// \param[in] r_coeff The coefficient of restitution. Must be in [0;1].
+  ///
+  /// \return A reference to the generalized velocity after impact stored in data.dq_after. The Lagrange Multipliers linked to the contact impulsed are available throw data.impulse_c vector.
+  ///
+  template<typename Scalar, int Options, template<typename,int> class JointCollectionTpl, typename ConfigVectorType, typename TangentVectorType, typename ConstraintMatrixType>
+  inline const typename DataTpl<Scalar,Options,JointCollectionTpl>::TangentVectorType &
+  impulseDynamics(const ModelTpl<Scalar,Options,JointCollectionTpl> & model,
+                  DataTpl<Scalar,Options,JointCollectionTpl> & data,
+                  const Eigen::MatrixBase<ConfigVectorType> & q,
+                  const Eigen::MatrixBase<TangentVectorType> & v_before,
+                  const Eigen::MatrixBase<ConstraintMatrixType> & J,
+                  const Scalar inv_damping,
+                  const Scalar r_coeff);
```

Comment:
I think a better signature would be: `impulseDynamics(model, data, q, v_before, J, r_coeff=0., inv_damping=0.)` ; i.e., switch `r_coeff` and `inv_damping`, so that `inv_damping` could be easily left unspecified.

Reply:
Also, this way calls like `impulseDynamics(model, data, q, v_before, J)` or `impulseDynamics(model, data, q, v_before, J, r_coeff)` would still be valid and return exactly the same result

Reply:
Yes, but 
`impulseDynamics(model, data, q, v_before, J, r_coeff=0., inv_damping=0.)`
has conflitcs with
`impulseDynamics(model, data, q, v_before, J, r_coeff=0., updateKinematics=true)`
that's why I resort to this solution, unfortunately.

Reply:
You can make it so that there are no conflicts
```
impulseDynamics(model, data, q, v_before, J, r_coeff=0., inv_damping=0.)
PINOCCHIO_DEPRECATED impulseDynamics(model, data, q, v_before, J, r_coeff, updateKinematics)
```
Without default arguments for the deprecated version.
In this way, if the user calls any of
```
impulseDynamics(model, data, q, v_before, J)
impulseDynamics(model, data, q, v_before, J, r_coeff)
```
The new version will be called, achieving the same effects as the old one. While the call
```
impulseDynamics(model, data, q, v_before, J, r_coeff, inv_damping)
```
will unambiguously refer to the new version and
```
impulseDynamics(model, data, q, v_before, J, r_coeff,false)
impulseDynamics(model, data, q, v_before, J, r_coeff,true)
```
will unambiguously refer to the old version

Reply:
I can handle it today. It will be fast to do during two meetings ;)
Thanks for this nice proposition.

Reply:
Ok, I leave it to you, then. If you want further feedback and/or work on the unit test, try to let me know before 4PM, because I will not be able to work after 5PM and after that I will be away for two weeks :)

Reply:
The Python unit test for contact dynamics already exists, so if you want to work on it you can just modify it: [unittest/python/bindings_dynamics.py](https://github.com/stack-of-tasks/pinocchio/blob/devel/unittest/python/bindings_dynamics.py)

Reply:
Sounds good.